doc: add the witness feature (#522)
harvester/docs: docs/advanced/witness.md

:::note

The current limitation is that the witness node can only be enabled when joining the cluster. Also, the cluster can only have one witness node.

:::

## Create Harvester cluster with witness node

When you create a new cluster, you can choose the witness role when joining the cluster.
For example, you create a cluster with `harvester-node-0`, `harvester-node-1` and `harvester-node-2`.

Choose the `harvester-node-1` as the witness node when joining the cluster. After the cluster is created, you would see the `harvester-node-1` only has the etcd capability as follows:

```
NAME↑               STATUS   ROLE                         VERSION               PODS     CPU      MEM    %CPU    %MEM    CPU/A    MEM/A AGE        
harvester-node-0    Ready    control-plane,etcd,master    v1.27.10+rke2r1         70    1095    10143      10      63    10000    15976 4d13h      
harvester-node-1    Ready    etcd                         v1.27.10+rke2r1          7     258     2258       2      14    10000    15976 4d13h      
harvester-node-2    Ready    control-plane,etcd,master    v1.27.10+rke2r1         36     840     6905       8      43    10000    15976 4d13h      
```

> You can see the `harvester-node-1` use less of resource to running as expection.

Reviewer: would be nice to add a note that the user needs to wait for the third node to join with a correct role and the controller can start the promotion procedure.

Author: updated!

## Workloads on the witness node

The witness node only runs the must workload like etcd. Now the following list is the supported workloads on the witness node.
- **harvester-node-manager**
- **cloud-controller-manager**
- **etcd**
- **kube-proxy**
- **rke2-canal**
- **rke2-multus**

## Upgrade the cluster with witness node

When you upgrade the cluster with a witness node, you can upgrade the cluster as usual. But if you have a degraded volume, the upgrade might fail. You can check the following limitations for the volume robustness.

## Limitations

There were some limitations for the cluster with witness.

> - Volume robustness: The witness node does not store any data. It means we will miss one replica when the cluster has the witness node due to the Longhorn default replica number (3). The upgrade might failed by the degraded volume.

Reviewer: ```suggestion
- Volume robustness: The witness node does not store any volume data. It means we will miss one replica when the cluster has the witness node due to the Longhorn default replica number (3). The upgrade might failed by the degraded volume.
```

It also happens when there are only 3 nodes in the cluster.

Author: updated!

  ![new storageclass replica 2](/img/v1.3/advanced/new-storageclass-rep-2.png)
  ![set to default](/img/v1.3/advanced/set-to-default-sc.png)

  If you already create some volume with the default StorageClass, you can manually decrease the replica counts from 3 -> 2 on Longhorn UI -> **Volume Page** as a workaround. You can get the related info for how to redirect to Longhorn UI from [here](../troubleshooting/harvester/#access-embedded-rancher-and-longhorn-dashboards)
  ![redirect-to-longhorn-volume-page](/img/v1.3/advanced/redirect-to-longhorn-vol-page.png)
  ![update-replica-count-to-2](/img/v1.3/advanced/update-replica-2.png)

  After, that the volume robustness will be healthy.

## Known issues

1. When you create a cluster with the witness node, the page of network config creation can not find any useable uplink with all nodes.
![create network config with all nodes](/img/v1.3/advanced/create-policy-with-all-nodes.png)
![no uplink](/img/v1.3/advanced/no-uplink.png)

  You can choose nodes other than the witness node as a workaround. Then, you can successfully create the network config.

>   ![create network config with specific node](/img/v1.3/advanced/create-policy-with-specific-node.png)

Reviewer: Please help add a note that the user needs to repeat this on all nodes that the network is supposed to span.

Author: updated!

---
sidebar_position: 8
sidebar_label: Witness Node involved Cluster

> title: "Witness Node involved Cluster"

Reviewer: How about just "Witness node"

Author: sounds ok to me, updated!

---

<head>
  <link rel="canonical" href="https://docs.harvesterhci.io/v1.1/advanced/witness/"/>
</head>

_Available as of v1.3.0_

On the edge cases, the witness node plays a specific role in maintaining the cluster's high availability. The witness node is a lightweight node that does not store any data and does not have any workload. This means the witness node can be a lightweight node, which will benefit both cost and high availability.

Harvester now supports the witness node involved cluster. You can enable the witness node when joining the cluster.

> For information about the role management, please refer to [role management](../TBD.md).

Author: I will update this link after https://github.com/harvester/docs/pull/520 is merged.

Author: updated!

:::note

The current limitation is that the witness node can only be enabled when joining the cluster. Also, the cluster can only have one witness node.

:::

## Create Harvester cluster with witness node

When you create a new cluster, you can choose the witness role when joining the cluster.
For example, you create a cluster with `harvester-node-0`, `harvester-node-1` and `harvester-node-2`.

Choose the `harvester-node-1` as the witness node when joining the cluster. After the cluster is created, you would see the `harvester-node-1` only has the etcd capability as follows:

```
NAME↑               STATUS   ROLE                         VERSION               PODS     CPU      MEM    %CPU    %MEM    CPU/A    MEM/A AGE        
harvester-node-0    Ready    control-plane,etcd,master    v1.27.10+rke2r1         70    1095    10143      10      63    10000    15976 4d13h      
harvester-node-1    Ready    etcd                         v1.27.10+rke2r1          7     258     2258       2      14    10000    15976 4d13h      
harvester-node-2    Ready    control-plane,etcd,master    v1.27.10+rke2r1         36     840     6905       8      43    10000    15976 4d13h      
```

> You can see the `harvester-node-1` use less of resource to running as expection. You need to satisfy the three-node requirement, and the promote controller will promote the second and third nodes.

Author: Hi @jillian-maroket, sorry to confuse you.

Here, I would like to express that we need to create a 3-node cluster and then promote the second and third nodes to witness nodes or management nodes, as an example.

> You can see the `harvester-node-1` use less of resource to running as expection. You need to satisfy the three-node requirement, and the promote controller will promote the second and third nodes.

Reviewer 2: Thanks. What I don't understand is why we are mentioning the three-node requirement and the need to promote nodes directly after the example. In any case, you can replace your sentence with something like this:

"Because the cluster must have three nodes, the promote controller will promote the other two nodes."

> You can see the `harvester-node-1` use less of resource to running as expection. You need to satisfy the three-node requirement, and the promote controller will promote the second and third nodes.

Author: Thanks!
I rewrote this part. Please help recheck it!

## Workloads on the witness node

The witness node only runs the must workload like etcd. Now the following list is the supported workloads on the witness node.
- **harvester-node-manager**
- **cloud-controller-manager**
- **etcd**
- **kube-proxy**
- **rke2-canal**
- **rke2-multus**

## Upgrade the cluster with witness node

When you upgrade the cluster with a witness node, you can upgrade the cluster as usual. But if you have a degraded volume, the upgrade might fail. You can check the following limitations for the volume robustness.

## Limitations

> There were some limitations for the cluster with witness.
> 
> - Volume robustness: The witness node does not store any volume data. It means we will miss one replica when the cluster has the witness node due to the Longhorn default replica number (3). The upgrade might failed by the degraded volume. It happens when the cluster only has three nodes, one of which is the witness node.
> 
>   To resolve this, you should create a new StorageClass and decrease its replica number from 3 -> 2. Then, set this StorageClass as a default StorageClass.
>   ![new storageclass replica 2](/img/v1.3/advanced/new-storageclass-rep-2.png)
>   ![set to default](/img/v1.3/advanced/set-to-default-sc.png)
> 
>   If you already create some volume with the default StorageClass, you can manually decrease the replica counts from 3 -> 2 on Longhorn UI -> **Volume Page** as a workaround. You can get the related info for how to redirect to Longhorn UI from [here](../troubleshooting/harvester/#access-embedded-rancher-and-longhorn-dashboards)
>   ![redirect-to-longhorn-volume-page](/img/v1.3/advanced/redirect-to-longhorn-vol-page.png)
>   ![update-replica-count-to-2](/img/v1.3/advanced/update-replica-2.png)
> 
>   After, that the volume robustness will be healthy.
> 

Reviewer 2: ```suggestion
The witness node does not store any data so only two replicas are created for each Longhorn volume in the cluster. However, the default StorageClass `harvester-longhorn` has a replica count value of `3` for high availability. If you use this StorageClass to create volumes, Longhorn is unable to create the configured number of replicas. This results in volumes being marked as **Degraded** on the Longhorn UI.

To resolve this issue, create a new *default* StorageClass with the **Number of Replicas** parameter set to **2**. This ensures that only two replicas are created for each Longhorn volume.
  
![new storageclass replica 2](/img/v1.3/advanced/new-storageclass-rep-2.png)
![set to default](/img/v1.3/advanced/set-to-default-sc.png)

If you already created volumes using the original default StorageClass, you can modify the replica count on the **Volume** screen of the [embedded Longhorn UI](../troubleshooting/harvester/#access-embedded-rancher-and-longhorn-dashboards).

![redirect-to-longhorn-volume-page](/img/v1.3/advanced/redirect-to-longhorn-vol-page.png)
![update-replica-count-to-2](/img/v1.3/advanced/update-replica-2.png)
```

Reviewer 2: A list is not necessary because there is only one limitation. I also removed "Volume robustness" because it distracts from the main point, which is the incorrect replica count.

> ## Known issues
> 
> 1. When you create a cluster with the witness node, the page of network config creation can not find any useable uplink with all nodes.
> ![create network config with all nodes](/img/v1.3/advanced/create-policy-with-all-nodes.png)
> ![no uplink](/img/v1.3/advanced/no-uplink.png)
> 
>   You can choose nodes other than the witness node as a workaround. Then, you can successfully create the network config.
>   ![create network config with specific node](/img/v1.3/advanced/create-policy-with-specific-node.png)
>   ![get uplink](/img/v1.3/advanced/get-uplink.png)
> 
>   You need to repeat this operation for every node except the witness node. The configuration for all nodes could be the same.
> 
>   Related issue: [[BUG] Unable to select NIC to create network config when cluster contains witness node](https://github.com/harvester/harvester/issues/5325)

Reviewer 2: ```suggestion
## Known Issues

When creating a cluster with a witness node, the **Network Config: Create** screen on the Harvester UI is unable to identify any NICs that can be used with all nodes.

![create network config with all nodes](/img/v1.3/advanced/create-policy-with-all-nodes.png)
![no uplink](/img/v1.3/advanced/no-uplink.png)

The workaround is to select a non-witness node and then select a NIC that can be used with that specific node.

![create network config with specific node](/img/v1.3/advanced/create-policy-with-specific-node.png)
![get uplink](/img/v1.3/advanced/get-uplink.png)

Related issue: [[BUG] Unable to select NIC to create network config when cluster contains witness node](https://github.com/harvester/harvester/issues/5325)
```

Author: Hi @jillian-maroket, 

This sentence
```
You need to repeat this operation for every node except the witness node. The configuration for all nodes could be the same.
```

We want to inform the user that the workaround must be executed on every non-witness node. How could we describe it here? Thanks!

---
sidebar_position: 8
sidebar_label: Witness Node
title: "Witness Node"
---

<head>
  <link rel="canonical" href="https://docs.harvesterhci.io/v1.1/advanced/witness/"/>
</head>

_Available as of v1.3.0_

> In edge cases, the witness node is responsible for establishing a member quorum within an etcd cluster and maintaining high availability. An etcd cluster requires a quorum (a majority of nodes) to agree on updates to the cluster state.
> 
> Witness nodes do not store any data, but the [hardware recommendations](https://etcd.io/docs/v3.3/op-guide/hardware/) for etcd nodes must still be considered. Using hardware with limited resources significantly affects cluster performance, as described in the article [Slow etcd performance (performance testing and optimization)](https://www.suse.com/support/kb/doc/?id=000020100). 

Reviewer: ```suggestion
A production Harvester cluster has a control plane with three management nodes (or control plane nodes). Each node runs an etcd service to act as a member in Kubernetes' etcd cluster. The node count of the control plane nodes has to be an odd number (it's always 3 in Harvester) to maintain etcd quorum.

However, there are cases in which a user doesn't want a management node to run workloads, or the user wants to create a two-node HA cluster. That's the situation in a Witness node comes into play.

A witness node doesn't run user workload or store any VM data, it's pure purpose is to maintain etcd's quorum. In general, it's OK to use a machine with a lower hardware spec, but the hardware still needs to satisfy etcd's [hardware recommendations](https://etcd.io/docs/v3.3/op-guide/hardware/). A hardware with limited resources significantly affects cluster performance, as described in the article [Slow etcd performance (performance testing and optimization)](https://www.suse.com/support/kb/doc/?id=000020100). 
```

My suggestion is we need to briefly introduce the control plane and then move to witness node role. I add some thoughts, can you see if this works.

Author: Thanks for the suggestions. Updated with minor changes.

---
sidebar_position: 8
sidebar_label: Witness Node
title: "Witness Node"
---

<head>
  <link rel="canonical" href="https://docs.harvesterhci.io/v1.1/advanced/witness/"/>
</head>

_Available as of v1.3.0_

> A production Harvester cluster needs control plane to manage nodes and pods. In Harvester, usually with three node management nodes, each node contains the whole control plane components.The etcd is one of the important parts of the control plane. Also, it is used by Kubernetes to store all of its data, including its configuration details, state, and metadata. The etcd nodes have to be an odd number (it's always 3 in Harvester) to maintain the etcd quorum.
> 
> Sometimes, a user wants to avoid a management node deploying any workload or user data. Each node runs an etcd service to act as a Kubernetes' etcd cluster member. That's the situation in a Witness node comes into play.

Reviewer 2: ```suggestion
Harvester clusters deployed in production environments require a control plane for node and pod management. A typical three-node cluster has three management nodes that each contain the complete set of control plane components. One key component is etcd, which Kubernetes uses to store its data (configuration, state, and metadata). The etcd node count must always be an odd number (for example, 3 is the default count in Harvester) to ensure that a quorum is maintained.

Some situations may require you to avoid deploying workloads and user data to management nodes. In these situations, one cluster node can be assigned the *witness* role, which limits it to functioning as an etcd cluster member. The witness node is responsible for establishing a member quorum (a majority of nodes), which must agree on updates to the cluster state.
```

Reviewer 2: These two paragraphs are new so I rewrote them.

> A witness node doesn't run user workload or store any VM data, it's pure purpose is to maintain etcd's quorum. In general, it's OK to use a machine with a lower hardware spec, but the hardware still needs to satisfy etcd's [hardware recommendations](https://etcd.io/docs/v3.3/op-guide/hardware/). A hardware with limited resources significantly affects cluster performance, as described in the article [Slow etcd performance (performance testing and optimization)](https://www.suse.com/support/kb/doc/?id=000020100). 

Reviewer 2: ```suggestion
Witness nodes do not store any data, but the [hardware recommendations](https://etcd.io/docs/v3.3/op-guide/hardware/) for etcd nodes must still be considered. Using hardware with limited resources significantly affects cluster performance, as described in the article [Slow etcd performance (performance testing and optimization)](https://www.suse.com/support/kb/doc/?id=000020100). 
```

Reviewer 2: I reviewed this previously but the change was not applied.
